### DISCUSSION TITLE
Unexpected Error: AssertionError("No characteristic with ['notify', 'indicate'] property found!")

### DISCUSSION THREAD

**Discussion by rogermeier** [Jakeler/ble-serial#69] [General]
I'm trying to connect to a [Bluefruit LE UART friend](https://www.adafruit.com/product/2479) module from Adafruit, on macOS BigSur.

ble-scan finds the device:

`6742AA30-3121-4A86-959D-6DFA5F6467DD (RSSI=-74): Flip Dot Clock`

Here are the results of the deep scan:

```
Started deep scan of 6742AA30-3121-4A86-959D-6DFA5F6467DD

Found device 6742AA30-3121-4A86-959D-6DFA5F6467DD: Flip Dot Clock (out of 1)
/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/ble_serial/scan/main.py:45: FutureWarning: This method will be removed future version, use the services property instead.
  print_details(await client.get_services())
SERVICE 00001530-1212-efde-1523-785feabcd123 (Handle: 12): Device Firmware Update Service
     CHARACTERISTIC 00001532-1212-efde-1523-785feabcd123 (Handle: 13): DFU Packet ['write-without-response']
     CHARACTERISTIC 00001531-1212-efde-1523-785feabcd123 (Handle: 15): DFU Control Point ['write', 'notify']
         DESCRIPTOR 00002902-0000-1000-8000-00805f9b34fb (Handle: 17): Client Characteristic Configuration
     CHARACTERISTIC 00001534-1212-efde-1523-785feabcd123 (Handle: 18): DFU Version ['read']
SERVICE 0000180a-0000-1000-8000-00805f9b34fb (Handle: 20): Device Information
     CHARACTERISTIC 00002a29-0000-1000-8000-00805f9b34fb (Handle: 21): Manufacturer Name String ['read']
     CHARACTERISTIC 00002a24-0000-1000-8000-00805f9b34fb (Handle: 23): Model Number String ['read']
     CHARACTERISTIC 00002a28-0000-1000-8000-00805f9b34fb (Handle: 25): Software Revision String ['read']
     CHARACTERISTIC 00002a26-0000-1000-8000-00805f9b34fb (Handle: 27): Firmware Revision String ['read']
     CHARACTERISTIC 00002a27-0000-1000-8000-00805f9b34fb (Handle: 29): Hardware Revision String ['read']
SERVICE 6e400001-b5a3-f393-e0a9-e50e24dcca9e (Handle: 31): Nordic UART Service
     CHARACTERISTIC 6e400003-b5a3-f393-e0a9-e50e24dcca9e (Handle: 32): Nordic UART TX ['notify']
         DESCRIPTOR 00002901-0000-1000-8000-00805f9b34fb (Handle: 34): Characteristic User Description
         DESCRIPTOR 00002902-0000-1000-8000-00805f9b34fb (Handle: 35): Client Characteristic Configuration
     CHARACTERISTIC 6e400002-b5a3-f393-e0a9-e50e24dcca9e (Handle: 36): Nordic UART RX ['write-without-response', 'write']
         DESCRIPTOR 00002901-0000-1000-8000-00805f9b34fb (Handle: 38): Characteristic User Description

Completed deep scan of 6742AA30-3121-4A86-959D-6DFA5F6467DD
```

I tried connecting with just the device UUID, and ble-serial asked me to specify -w and -r because it says it could find those characteristics (even though they seem to be present in the results of the deep scan).

I then tried to connect like this:

```
ble-serial -t 10 -d -v 6742AA30-3121-4A86-959D-6DFA5F6467DD -w 6e400002-b5a3-f393-e0a9-e50e24dcca9e -r 6e400002-b5a3-f393-e0a9-e50e24dcca9e

18:46:20.372 | DEBUG | main.py: Running: Namespace(verbose=1, port='/tmp/ttyBLE', timeout=10.0, adapter='hci0', mtu=20, device='6742AA30-3121-4A86-959D-6DFA5F6467DD', addr_type='public', service_uuid=None, write_uuid='6e400002-b5a3-f393-e0a9-e50e24dcca9e', read_uuid='6e400002-b5a3-f393-e0a9-e50e24dcca9e', mode='rw', filename=None, binlog=False)
18:46:20.372 | DEBUG | selector_events.py: Using selector: KqueueSelector
18:46:20.374 | INFO | linux_pty.py: Slave created on /tmp/ttyBLE -> /dev/ttys001
18:46:20.374 | INFO | ble_interface.py: Receiver set up
/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/ble_serial/bluetooth/ble_interface.py:28: FutureWarning: This method will be removed future version, pass the callback to the BleakClient constructor instead.
  self.dev.set_disconnected_callback(self.handle_disconnect)
18:46:27.742 | INFO | ble_interface.py: Trying to connect with 6742AA30-3121-4A86-959D-6DFA5F6467DD: Flip Dot Clock
18:46:29.287 | INFO | ble_interface.py: Device 6742AA30-3121-4A86-959D-6DFA5F6467DD connected
18:46:29.287 | DEBUG | ble_interface.py: Characteristic candidates for write: 
	6e400002-b5a3-f393-e0a9-e50e24dcca9e (Handle: 36): Nordic UART RX ['write-without-response', 'write']
18:46:29.287 | INFO | ble_interface.py: Found write characteristic 6e400002-b5a3-f393-e0a9-e50e24dcca9e (H. 36)
18:46:29.287 | DEBUG | ble_interface.py: Characteristic candidates for notify: 
	6e400002-b5a3-f393-e0a9-e50e24dcca9e (Handle: 36): Nordic UART RX ['write-without-response', 'write']
18:46:29.287 | ERROR | main.py: Unexpected Error: AssertionError("No characteristic with ['notify', 'indicate'] property found!")
18:46:29.287 | WARNING | main.py: Shutdown initiated
18:46:29.288 | INFO | linux_pty.py: Serial reader and symlink removed
18:46:29.323 | WARNING | ble_interface.py: Device 6742AA30-3121-4A86-959D-6DFA5F6467DD disconnected
18:46:29.323 | INFO | ble_interface.py: Stopping Bluetooth event loop
18:46:29.323 | INFO | ble_interface.py: Bluetooth disconnected
18:46:29.324 | INFO | main.py: Shutdown complete.
```

Am I doing something wrong? I know the device is operational as I can connect to and communicate with it with Adafruit's Bluefruit Connect tool just fine.

**Comment by rogermeier** [comment]
I figured it out. I mistakenly pasted the wrong ID for the -w argument (hard to spot since they only differ in a single digit)

This works:

`ble-serial -t 10 -v -d 6742AA30-3121-4A86-959D-6DFA5F6467DD -w 6e400002-b5a3-f393-e0a9-e50e24dcca9e -r 6e400003-b5a3-f393-e0a9-e50e24dcca9e`

Very nice. I can now connect to my module with 'screen' on the Mac.

**Reply by Jakeler** [reply]
Nice! 
Also I already added the known nordic UART profile to the autodetection list, this is just not in 2.6.0, but I will make a 2.6.1 release soon.
If you want to try it out immediately it can be installed with:
```
pip install https://github.com/Jakeler/ble-serial/archive/refs/heads/master.zip
```
Then that should just work:
```
ble-serial -t 10 -v -d 6742AA30-3121-4A86-959D-6DFA5F6467DD
```

**Reply by rogermeier** [reply]
Thanks! That's great! I have confirmed that the updated version works (I had to use --force-reinstall to get it to install over 2.6.0).

BTW, do you happen to know if there is a different way to create virtual ports on macOS and Linux? I'm the author of a serial terminal app ([CoolTerm](https://freeware.the-meiers.org#CoolTerm)), and the serial port API I'm using is not working with the virtual ports created by ble-serial (and also socat) and macOS and Linux. Attempting to connect to it results in an "Error 25" which translates to 'Inappropriate ioctl for device’, whatever that means.

**Reply by Jakeler** [reply]
Yes, on Linux you could try [tty0tty](https://github.com/freemed/tty0tty), comparable to com0com on Windows, but I am not aware of anything similar for macOS. 

`ioctl` calls are used to get/set attributes in the device driver, so settings affecting the hardware, for example baud rate. As far I can tell tty0tty implements at least the common ioctl command codes. Then socat could be used as bridge between the two port pairs.

The [pty](https://man7.org/linux/man-pages/man7/pty.7.html) used by ble-serial will probaly not implement quite the same interface as a [hardware tty](https://man7.org/linux/man-pages/man4/tty_ioctl.4.html). If the library used the slightly higher level [termios](https://man7.org/linux/man-pages/man3/termios.3.html) I would expect it to work though. Pretty much all serial tools just worked, so I did not look too deep into it. It could be interesting to run you terminal app with `strace` to figure out which syscall exactly fails.

**Reply by rogermeier** [reply]
Thanks. I've been digging a little deeper. It doesn't seem like the error occurs at the moment when the connection is made but sometime afterwards, possibly when port setting are applied. I'll see if I can pinpoint exactly when.
My app automatically disconnect when the "Error" event of the serial port control fires. If I can catch exactly when the error occurs, then I might be able to prevent the port from closing automatically.

**Reply by rogermeier** [reply]
I have found the exact call to the API that causes the error. To activate any port settings, the 'Reset' method of the serial port control needs to be called. That's when the error is raised. Unfortunately, I have no idea what happens exactly under the hood when "Reset" is called, but it is when things like baudrate, databits, stop bits, flowcontrol, etc is applied. I have no way of telling which setting is the offending one since I have to assume they are all set during a call to "Reset".

**Reply by rogermeier** [reply]
I've been able to get CoolTerm working with ble-serial by simply having the code ignoring any 'Inappropriate ioctl for device’ errors and opening the port anyways. This seems to work well and I haven't had any issues communicating with my BlueFruit LE module. I will be releasing the next version of CoolTerm with this workaround.
